### PR TITLE
Investigate adding ErrorBoundry to ListView

### PR DESCRIPTION
**Pull request by jcaianirh**
https://jira.coreos.com/browse/CONSOLE-536

When a js issue occurs, prevent the white screen and provide an Oh no... message as well as output showing the output to the issue with the page along with available traces from the componentDidCatch Error Boundary class.

The component as of now is simple enough to work as a full page error, or an error that is part of other page content (see figures below) such as a list in a details view.

Spoke briefly with @bmignano and @spadgett about design, and figured we could discuss here for a more detailed dialog.

@openshift/team-ux-review 

screen shot 2018-11-06 at 2 12 22 pm

screen shot 2018-11-06 at 2 13 07 pm

screen shot 2018-11-06 at 3 20 58 pm

screen shot 2018-11-06 at 3 21 17 pm

**Updated 11/13**
screen shot 2018-11-13 at 10 59 34 am

### CODE REVIEW

**Comment by spadgett** [comment]
I'd recommend `copy-to-clipboard` for the stack so it's easy to copy. Consider `white-space: prewrap;` to preserve the formatting.

For details pages that fail, it would be nice to give a link to the YAML tab so you can see the content.

cc @alecmerdler

**Comment by cshinn** [comment]
@stacymcauliffe thoughts on the message text?

**Comment by bmignano** [comment]
I'm wondering whether there's the possibility of using something like PatternFly's [empty state](https://www.patternfly.org/pattern-library/communication/empty-state/) here. The pattern seems to exclude this type of error state, but PF doesn't provide a design specifically for error states. Seems like there might be potential here to start with an empty state and modify it to meet our needs. Thoughts?

File: frontend/public/components/utils/error-boundary.tsx [hunk 1]
```diff
@@ -30,12 +46,14 @@ export const withFallback: WithFallback = (Component, FallbackComponent) => (pro
   <Component {...props} />
 </ErrorBoundary>;
 
-export type WithFallback = <P = {}>(Component: React.ComponentType<P>, FallbackComponent?: React.ComponentType) => React.ComponentType<P>;
+export type WithFallback = <P = {}>(Component: React.ComponentType<P>, FallbackComponent?: React.ComponentType<ErrorBoundaryFallbackComponentProps>) => React.ComponentType<P>;
 
 export type ErrorBoundaryProps = {
-  FallbackComponent?: React.ComponentType;
+  FallbackComponent?: React.ComponentType<ErrorBoundaryFallbackComponentProps>;
```
Comment (alecmerdler) [inline frontend/public/components/utils/error-boundary.tsx]: We are going to require that *every* fallback component uses these props?

Comment (jcaianirh) [reply frontend/public/components/utils/error-boundary.tsx]: @alecmerdler probably not a good idea, so React.ComponentType<any> to broaden that restriction, right?

Comment (jcaianirh) [reply frontend/public/components/utils/error-boundary.tsx]: @alecmerdler oh yes. This is all the possible data that a Component could use to react to the Error Boundary.

**Comment by stacymcauliffe** [comment]
@bmignano @cshinn - RE: the empty state pattern, I totally agree. When we treat error states as empty states we can use them to accomplish the following things:
-Give the user a very brief overview statement of what's gone wrong or what is happening  
-Provide the error code if one is available
-Provide a bit more context and info about how the user can move forward
-Provide CTAs and/or links out to related content or docs

The idea is to help the user understand and react very quickly. 

The only gotcha with this is that we'd need to do some work to create error messages that are less generic than "Oh no! Something went wrong." 

If we needed to keep it super generic (which I wouldnt recommend if we can help it bc it doesn't help users solve the problem) then I'd consider getting rid of the "Oh no!" and just keeping "Something went wrong." because "Oh no!" feels alarming to me. 

Here's an example of one we did for OSIO:
image

**Comment by spadgett** [comment]
Thanks @stacymcauliffe.

For context, this is an error we display specifically when we have an uncaught exception in the code that would otherwise crash the app (blank white page, no way to recover). There's not a lot of context we can show except for the stack trace detail that @jcaianirh has above in the screenshot. The user should not see unless we have a bug.

There's not much the user can do in these cases except go back to the previous page or report the problem.

**Comment by stacymcauliffe** [comment]
Thanks @spadgett! Gotcha. I think what Joe is proposing works for what you're describing. I'd still clip the message to "Something went wrong." and maybe add another line of below that provides link text to "Go back" or "Report the problem"

File: frontend/public/components/factory/list-page.jsx [hunk 2]
```diff
@@ -350,6 +353,7 @@ export const ListPage = props => {
 };
 
 ListPage.displayName = 'ListPage';
+export const ListPageWrapper = listPageWithFallback(ListPage);
```
Comment (alecmerdler) [inline frontend/public/components/factory/list-page.jsx]: I don't know that we need another `ListPage` wrapper.

Comment (jcaianirh) [reply frontend/public/components/factory/list-page.jsx]: removed the wrapper and moved the hoc to the list page const declaration

File: frontend/public/components/utils/error-boundary.tsx [hunk 3]
```diff
@@ -11,17 +12,32 @@ class DefaultFallback extends React.Component {
...
+      },
+      errorInfo: {
+        componentStack: ''
+      }
+    };
   }
 
-  componentDidCatch(error) {
-    this.setState({hasError: true});
+  componentDidCatch(error, errorInfo) {
```
[the remark this answers is not in the document]
Comment (spadgett) [reply frontend/public/components/utils/error-boundary.tsx]: > This is not supported in our rev of react yet.

We should consider updating if it's not too hard

[the remark this answers is not in the document]
Comment (alecmerdler) [reply frontend/public/components/utils/error-boundary.tsx]: We've upgraded to `react@16.6.3`; but `getDerivedStateFromError()` doesn't give you the full `errorInfo`, so maybe we keep using `componentDidCatch()`.

**Pull request merged**